controllers/metal3.io/host_state_machine.go

```diff
@@ -226,6 +230,34 @@ func (hsm *hostStateMachine) checkInitiateDelete() bool {
 	return true
 }
 
+func (hsm *hostStateMachine) checkDeregisteredHost(info *reconcileInfo) (result actionResult) {
+	// If the deregistered annotation is set we remove any host from the
+	// provisioner and take no further action
+	// Note this doesn't change the current state, only the OperationalStatus
+	annotations := hsm.Host.GetAnnotations()
+	if annotations != nil {
+		if _, ok := annotations[metal3v1alpha1.DeregisteredAnnotation]; ok {
+			if info.host.OperationalStatus() != metal3v1alpha1.OperationalStatusDeregistered {
+				_, err := hsm.Provisioner.Delete()
```

Comment:
> @hardys That depends on amount of hosts, connection and memory limits in mysql db, on big clusters this will be an issue. One would have to tweak mysql config.

Right, but that's not what the provisioningCapacity slots are for AFAIK - those exist to prevent "thundering herd" issues where you try to simultaneously provision $many hosts, which results multiple ironic API calls and related conductor processing, DoSing the PXE infrastructure, saturating the physical network fabric, etc

In this case the annotation results in exactly one API call to Ironic, and it sounds like the consensus is to just make the call - if we see DB/ironic-api bottlenecks at scale we could revisit the discussion and/or look at ways to improve the Ironic/DB deployment configuration

Reply:
Moved this logic into `BareMetalHostReconciler`